File: packages/sdks/everstack-python/src/everstack/resources/evaluations.py
"""Evaluations resource."""

from __future__ import annotations

from typing import Any, Dict

from .._transport import Transport, AsyncTransport


class _Runs:
    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def create(self, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request("POST", "/v1/eval-runs", json_body=kwargs)

    def get(self, run_id: str) -> Dict[str, Any]:
        return self._t.request("GET", f"/v1/eval-runs/{run_id}")

    def list(self, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request("GET", "/v1/eval-runs", params=kwargs)

    def cancel(self, run_id: str) -> Dict[str, Any]:
        return self._t.request("POST", f"/v1/eval-runs/{run_id}/cancel", json_body={})

    def delete(self, run_id: str) -> Dict[str, Any]:
        return self._t.request("DELETE", f"/v1/eval-runs/{run_id}")

    def retry(self, run_id: str) -> Dict[str, Any]:
        return self._t.request("POST", f"/v1/eval-runs/{run_id}/retry", json_body={})

    def get_items(self, eval_run_id: str, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request("GET", f"/v1/eval-runs/{eval_run_id}/items", params=kwargs)

    def get_summary(self, run_id: str) -> Dict[str, Any]:
        return self._t.request("GET", f"/v1/eval-runs/{run_id}/summary")

    def compare(self, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request("POST", "/v1/eval-runs/compare", json_body=kwargs)

    def set_baseline(self, eval_run_id: str, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request(
            "POST", f"/v1/eval-runs/{eval_run_id}/baseline", json_body=kwargs
        )


class _Schedules:
    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def create(self, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request("POST", "/v1/eval-schedules", json_body=kwargs)

    def get(self, schedule_id: str) -> Dict[str, Any]:
        return self._t.request("GET", f"/v1/eval-schedules/{schedule_id}")

    def list(self, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request("GET", "/v1/eval-schedules", params=kwargs)

    def update(self, schedule_id: str, **kwargs: Any) -> Dict[str, Any]:
        return self._t.request(
            "PATCH", f"/v1/eval-schedules/{schedule_id}", json_body=kwargs
        )

    def delete(self, schedule_id: str) -> Dict[str, Any]:
        return self._t.request("DELETE", f"/v1/eval-schedules/{schedule_id}")


class Evaluations:
    """Sync evaluations resource."""

    runs: _Runs
    schedules: _Schedules

    def __init__(self, transport: Transport) -> None:
        self._t = transport
        self.runs = _Runs(transport)
        self.schedules = _Schedules(transport)

    def score(
        self,
        *,
        input: Any = None,
        output: Any = None,
        expected_output: Any = None,
        metadata: Any = None,
        scorer_config_ids: Any = None,
    ) -> Dict[str, Any]:
        """Synchronously score an output against one or more score configs
        (POST /v1/score-output). Powers the pytest-native ``everstack.evals``
        harness."""
        body = {
            "input": input,
            "output": output,
            "expectedOutput": expected_output,
            "metadata": metadata,
            "scorerConfigIds": scorer_config_ids or [],
        }
        body = {k: v for k, v in body.items() if v is not None}
        return self._t.request("POST", "/v1/score-output", json_body=body)


class AsyncEvaluations:
    """Async evaluations resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._t = transport

    async def create_run(self, **kwargs: Any) -> Dict[str, Any]:
        return await self._t.request("POST", "/v1/eval-runs", json_body=kwargs)

    async def get_run(self, run_id: str) -> Dict[str, Any]:
        return await self._t.request("GET", f"/v1/eval-runs/{run_id}")

    async def list_runs(self, **kwargs: Any) -> Dict[str, Any]:
        return await self._t.request("GET", "/v1/eval-runs", params=kwargs)

    async def cancel_run(self, run_id: str) -> Dict[str, Any]:
        return await self._t.request("POST", f"/v1/eval-runs/{run_id}/cancel", json_body={})

    async def get_run_summary(self, run_id: str) -> Dict[str, Any]:
        return await self._t.request("GET", f"/v1/eval-runs/{run_id}/summary")

File: packages/sdks/everstack-go/evaluations.go
package everstack

import (
	"context"
	"fmt"
	"net/url"
)

// EvaluationsResource provides evaluation operations.
type EvaluationsResource struct {
	Runs      *EvalRunsResource
	Schedules *EvalSchedulesResource
}

func newEvaluationsResource(t *Transport) *EvaluationsResource {
	return &EvaluationsResource{
		Runs:      &EvalRunsResource{t: t},
		Schedules: &EvalSchedulesResource{t: t},
	}
}

// --- Runs ---

// EvalRunsResource provides eval run operations.
type EvalRunsResource struct {
	t *Transport
}

// Create creates an eval run.
func (r *EvalRunsResource) Create(ctx context.Context, body map[string]any) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "POST", "/v1/eval-runs", body, nil, &resp)
}

// Get retrieves an eval run by ID.
func (r *EvalRunsResource) Get(ctx context.Context, runID string) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "GET", fmt.Sprintf("/v1/eval-runs/%s", runID), nil, nil, &resp)
}

// List lists eval runs.
func (r *EvalRunsResource) List(ctx context.Context, params ...url.Values) (map[string]any, error) {
	var q url.Values
	if len(params) > 0 {
		q = params[0]
	}
	var resp map[string]any
	return resp, r.t.Request(ctx, "GET", "/v1/eval-runs", nil, q, &resp)
}

// Cancel cancels a running eval.
func (r *EvalRunsResource) Cancel(ctx context.Context, runID string) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "POST", fmt.Sprintf("/v1/eval-runs/%s/cancel", runID), map[string]any{}, nil, &resp)
}

// Delete deletes an eval run.
func (r *EvalRunsResource) Delete(ctx context.Context, runID string) error {
	return r.t.Request(ctx, "DELETE", fmt.Sprintf("/v1/eval-runs/%s", runID), nil, nil, nil)
}

// Retry retries a failed eval run.
func (r *EvalRunsResource) Retry(ctx context.Context, runID string) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "POST", fmt.Sprintf("/v1/eval-runs/%s/retry", runID), map[string]any{}, nil, &resp)
}

// GetItems retrieves items for an eval run.
func (r *EvalRunsResource) GetItems(ctx context.Context, evalRunID string, params ...url.Values) (map[string]any, error) {
	var q url.Values
	if len(params) > 0 {
		q = params[0]
	}
	var resp map[string]any
	return resp, r.t.Request(ctx, "GET", fmt.Sprintf("/v1/eval-runs/%s/items", evalRunID), nil, q, &resp)
}

// GetSummary retrieves the summary for an eval run.
func (r *EvalRunsResource) GetSummary(ctx context.Context, runID string) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "GET", fmt.Sprintf("/v1/eval-runs/%s/summary", runID), nil, nil, &resp)
}

// Compare compares two eval runs.
func (r *EvalRunsResource) Compare(ctx context.Context, body map[string]any) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "POST", "/v1/eval-runs/compare", body, nil, &resp)
}

// SetBaseline sets a baseline for an eval run.
func (r *EvalRunsResource) SetBaseline(ctx context.Context, evalRunID string, body map[string]any) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "POST", fmt.Sprintf("/v1/eval-runs/%s/baseline", evalRunID), body, nil, &resp)
}

// --- Schedules ---

// EvalSchedulesResource provides eval schedule operations.
type EvalSchedulesResource struct {
	t *Transport
}

// Create creates an eval schedule.
func (r *EvalSchedulesResource) Create(ctx context.Context, body map[string]any) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "POST", "/v1/eval-schedules", body, nil, &resp)
}

// Get retrieves an eval schedule by ID.
func (r *EvalSchedulesResource) Get(ctx context.Context, scheduleID string) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "GET", fmt.Sprintf("/v1/eval-schedules/%s", scheduleID), nil, nil, &resp)
}

// List lists eval schedules.
func (r *EvalSchedulesResource) List(ctx context.Context, params ...url.Values) (map[string]any, error) {
	var q url.Values
	if len(params) > 0 {
		q = params[0]
	}
	var resp map[string]any
	return resp, r.t.Request(ctx, "GET", "/v1/eval-schedules", nil, q, &resp)
}

// Update updates an eval schedule.
func (r *EvalSchedulesResource) Update(ctx context.Context, scheduleID string, body map[string]any) (map[string]any, error) {
	var resp map[string]any
	return resp, r.t.Request(ctx, "PATCH", fmt.Sprintf("/v1/eval-schedules/%s", scheduleID), body, nil, &resp)
}

// Delete deletes an eval schedule.
func (r *EvalSchedulesResource) Delete(ctx context.Context, scheduleID string) error {
	return r.t.Request(ctx, "DELETE", fmt.Sprintf("/v1/eval-schedules/%s", scheduleID), nil, nil, nil)
}
